### PR TITLE
cp - filter the images by tag on the index page

### LINKED ISSUE #4: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [ ] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [ ] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by AgentCyrusJr**
Closes #4 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -17,7 +17,14 @@ def create
...
 
   def index
-    @list_of_images = Image.order(created_at: :desc)
+    if params[:tag_param].nil?
+      @list_of_images = Image.order(created_at: :desc)
+    else
+      @tag = params[:tag_param]
+      @list_of_images = Image.order(created_at: :desc).select do |image|
+        image.tag_list.include? @tag
+      end
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: You could use `Image.tagged_with(@tag)` to find those images with this tag.

Comment (shiboying) [reply app/controllers/images_controller.rb]: And if `Image.tagged_with(@tag)` returns empty, it means the `@tag` is invalid. I recommend you could use a flash message to display customer-facing message.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -17,7 +17,14 @@ def create
...
+    end
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: Also, you could extra the above logic into something called `ImageTagService` under `app/services` so instead you could call method from the service directly, e.g. `ImageTagService. images_with_tag`.

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -3,7 +3,7 @@
 <% @list_of_images.each do | image | %>
     <%= image_tag image.url, class:'saved-image' %>
     <% image.tag_list.each do |tag| %>
-        <%= tag %>
+        <%= link_to tag, root_path(tag_param: tag), class: "btn btn-primary"%>
```
Comment (shiboying) [inline app/views/images/index.html.erb]: for `class: "btn btn-primary"` we usually use single quote: `class: 'btn btn-primary'`

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -63,4 +63,18 @@ def test_image_seeds
...
+    assert_response :ok
+    assert_select '.saved-image', 1
+  end
+
+  def test_index__no_such_tag
+    get root_path(tag_param: 'no_such_tag')
+
+    assert_response :ok
+    assert_select '.saved-image', 0
+  end
```
Comment (shiboying) [inline test/controllers/images_controller_test.rb]: You could also add a test case for `tag_param` is `nil`.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -17,7 +17,14 @@ def create
   end
 
   def index
-    @list_of_images = Image.order(created_at: :desc)
+    if params[:tag_param].nil?
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: As `tag_param` is one of the `params`, I recommend you rename it to just `tag`.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -17,7 +19,13 @@ def create
   end
 
   def index
-    @list_of_images = Image.order(created_at: :desc)
+    if params[:tag].nil?
+      @list_of_images = Image.order(created_at: :desc)
+    else
+      @tag = params[:tag]
+      @list_of_images = ImagesFilteringService.run(tag: @tag)
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: It seems you do not use `@tag` in the `index` view.  Instead of create a instant variable, you could use `tag = params[:tag]`.

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -1,3 +1,5 @@
+require_relative '../services/images_filtering_service'
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: I think you could remove this line.
